src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+    pub indent: u16,
+}
+
+impl<T, U> Node<T, U> {
+    /// Create a new node, with data, but without a parent or children.
+    pub fn new(data: T, ty: U, indent: u16) -> Node<T, U> {
+        Node {
+            parent: None,
+            previous_sibling: None,
+            next_sibling: None,
+            first_child: None,
+            last_child: None,
+            data: data,
+            ty: ty,
+            indent: indent,
+        }
+    }
+
+    /// `true` if this node has no children.
+    pub fn is_leaf(&self) -> bool {
+        match self.first_child {
+            Some(_) => false,
+            None => true,
+        }
+    }
+
+    /// `true` if this node has no parent.
+    pub fn is_root(&self) -> bool {
+        match self.parent {
+            Some(_) => false,
+            None => true,
+        }
+    }
+
+    /// Return the ID of the parent node, if there is one.
+    pub fn parent(&self) -> Option<NodeId> {
+        self.parent
+    }
+
+    /// Return the ID of the first child of this node, if there is one.
+    pub fn first_child(&self) -> Option<NodeId> {
+        self.first_child
+    }
+
+    /// Return the ID of the last child of this node, if there is one.
+    pub fn last_child(&self) -> Option<NodeId> {
+        self.last_child
+    }
+
+    /// Return the ID of the previous sibling of this node, if there is one.
+    pub fn previous_sibling(&self) -> Option<NodeId> {
+        self.previous_sibling
+    }
+
+    /// Return the ID of the previous sibling of this node, if there is one.
+    pub fn next_sibling(&self) -> Option<NodeId> {
+        self.next_sibling
+    }
+}
+
+impl<T, U> fmt::Display for Node<T, U>
+    where T: fmt::Display,
+          U: fmt::Display
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut s = String::new();
+        for _ in 0..self.indent {
+            s.push_str(" ");
+        }
+        write!(f, "{}", s)?; // Indentation.
+        let data_s = format!(" {}", self.data);
+        if data_s.len() == 1 {
+            return write!(f, "{}", self.ty);
+        }
+        write!(f, "{}", self.ty)?;
+        write!(f, "{}", data_s)
+    }
+}
+
+// Turn a grammar, parser and input string into an AST arena.
+fn parse_into_ast(pt: &parser::Node<u16>, grm: &Grammar, input: &str) -> Arena<String, String> {
+    let mut arena = Arena::new();
+    let mut st = vec![(0, pt)]; // Stack of (indent level, node) pairs
+    // Stack of Option(NodeId)s which are parents of nodes on the `st` stack.
+    // The stack should never be empty, a `None` should be at the bottom of the stack.
+    let mut parent = vec![None];
+    let mut child_node: NodeId;
+    while !st.is_empty() {
+        let (indent, e) = st.pop().unwrap();
+        match e {
+            &parser::Node::Terminal { lexeme } => {
+                let tid: usize = lexeme.tok_id().try_into().ok().unwrap();
+                let tn = grm.term_name(TIdx::from(tid)).unwrap();
```

Comment:
I disagree: long variable names don't make code easier to read IMHO. The older I get, the shorter my variable names become, and the easier my code is to read.